### PR TITLE
Updated detailed report and results columns

### PR DESCRIPTION
**Pull request by aramesh7**
GPU detailed report
[detailed_report_with_info_columns.pdf](https://github.com/triton-inference-server/model_analyzer/files/6635283/detailed_report_with_info_columns.pdf)

CPU detailed report
[detailed_report_with_info_cpu.pdf](https://github.com/triton-inference-server/model_analyzer/files/6635102/detailed_report_with_info_cpu.pdf)

Here is an example of the new results columns
```
Models (Inference):
Model                       Batch   Concurrency   Model Config Path              Instance Group   Preferred Batch Sizes   Satisfies Constraints   Throughput (infer/sec)   p99 Latency (ms)   RAM Usage (MB)  
classification_malaria_v1   32      2             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     5248.0                   14.1               2889.0          
classification_malaria_v1   32      1             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     3744.0                   11.8               2872.0          
classification_malaria_v1   16      2             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     3056.0                   14.0               2870.0          
classification_malaria_v1   16      1             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     2592.0                   8.7                2868.0          

Models (GPU Metrics):
Model                       GPU UUID                                   Batch   Concurrency   Model Config Path              Instance Group   Preferred Batch Sizes   Satisfies Constraints   GPU Memory Usage (MB)   GPU Utilization (%)   GPU Power Usage (W)  
classification_malaria_v1   GPU-e35ba3d2-6eef-2bb9-e35c-6ef6eada4f11   32      2             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     1142.0                  32.6                  120.4                
classification_malaria_v1   GPU-e35ba3d2-6eef-2bb9-e35c-6ef6eada4f11   32      1             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     1142.0                  22.5                  105.8                
classification_malaria_v1   GPU-e35ba3d2-6eef-2bb9-e35c-6ef6eada4f11   16      2             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     1142.0                  20.4                  100.7                
classification_malaria_v1   GPU-e35ba3d2-6eef-2bb9-e35c-6ef6eada4f11   16      1             classification_malaria_v1_i0   1/GPU            [32]                    Yes                     1142.0                  18.9                  84.1                 

Server Only:
Model           GPU UUID                                   GPU Memory Usage (MB)   GPU Utilization (%)   GPU Power Usage (W)  
triton-server   GPU-e35ba3d2-6eef-2bb9-e35c-6ef6eada4f11   436.0                   0.0                   56.3                 
```

### CODE REVIEW

File: model_analyzer/reports/report_manager.py [hunk 1]
```diff
@@ -239,18 +243,32 @@ def _build_detailed_report(self, report_model_config):
...
+        detailed_info = self._build_detailed_info(report_key)
+        detailed_report.add_line_breaks(num_breaks=2)
+        detailed_report.add_paragraph(detailed_info, font_size=18)
+        detailed_report.add_paragraph(
+            "The first plot above shows the exact breakdown of the latencies in "
+            "the latency throughput curve for this model config. Following that "
+            "are the requested configurable plots showing the relationship between "
+            "various metrics measured by the Model Analyzer. The above table contains "
+            "detailed data for each of the measurements taken for this model config in "
+            "decreasing order.",
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:270]: In decreasing order of throughput?

File: model_analyzer/reports/report_manager.py [hunk 2]
```diff
@@ -239,18 +243,32 @@ def _build_detailed_report(self, report_model_config):
...
         # Next add table of measurements
         detailed_table = self._build_detailed_table(report_key)
         detailed_report.add_table(table=detailed_table)
 
+        # Add some details about the config
+        detailed_info = self._build_detailed_info(report_key)
+        detailed_report.add_line_breaks(num_breaks=2)
+        detailed_report.add_paragraph(detailed_info, font_size=18)
+        detailed_report.add_paragraph(
+            "The first plot above shows the exact breakdown of the latencies in "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:265]: Remove exact

File: model_analyzer/reports/report_manager.py [hunk 3]
```diff
@@ -265,9 +283,16 @@ def _build_summary_report(self, report_key, num_configs, statistics):
         total_configurations = statistics.total_configurations(report_key)
         num_best_configs = min(num_configs, total_configurations)
 
-        cpu_only, gpu_names, max_memories = self._get_gpu_stats(
-            report_key=report_key)
+        # Get GPU names and memory
+        model_config = self._summary_data[report_key][0][0]
+        cpu_only = model_config.cpu_only()
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:288]: Can you share a CPU only version of the report too?

File: model_analyzer/reports/report_manager.py [hunk 4]
```diff
@@ -467,63 +495,96 @@ def _build_detailed_table(self, model_config_name):
...
+        model_config, measurements = self._detailed_report_data[
+            model_config_name]
+        instance_group_string = model_config.instance_group_string()
+        dynamic_batching_string = model_config.dynamic_batching_string()
+        platform = model_config.get_field('platform')
+
+        # Measurements and GPU info
+        if model_config.cpu_only():
+            sentence = (
+                f"The model config {model_config_name} requests {instance_group_string} "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:548]: "The model config {model_config_name}  uses {n} {KIND} instances"

File: model_analyzer/reports/report_manager.py [hunk 5]
```diff
@@ -467,63 +495,96 @@ def _build_detailed_table(self, model_config_name):
...
+                f"instances. {len(measurements)} measurements were obtained for the model config "
+                f"on CPU with platfrom {platform}. ")
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:550]: {platform} really is not related to CPU. I think we should not include the {platform}. It is not something that we sweep on.

File: model_analyzer/reports/report_manager.py [hunk 6]
```diff
@@ -467,63 +495,96 @@ def _build_detailed_table(self, model_config_name):
...
+        else:
+            gpu_dict = self._get_gpu_stats(measurements=measurements)
+            gpu_names = ','.join(list(gpu_dict.keys()))
+            max_memories = ','.join(
+                [str(x) + ' GB' for x in gpu_dict.values()])
+            sentence = (
+                f"The model config {model_config_name} requests {instance_group_string} "
+                f"instances. {len(measurements)} measurements were obtained for the model config "
+                f"on GPU(s) {gpu_names} with memory limit(s) {max_memories}. This model "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:559]: How does this paragraph look like on a multi-GPU machine?

File: model_analyzer/reports/report_manager.py [hunk 7]
```diff
@@ -467,63 +495,96 @@ def _build_detailed_table(self, model_config_name):
...
+                f"uses the platform {platform}. ")
+
+        # Dynamic batching
+        if dynamic_batching_string == 'N/A':
+            sentence += "This model does not support batching. "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:564]: How do we know that it doesn't support dynamic batching? Is it for the case that `max_batch_size=0`?

File: model_analyzer/reports/report_manager.py [hunk 8]
```diff
@@ -467,63 +495,96 @@ def _build_detailed_table(self, model_config_name):
...
+                f"The model config {model_config_name} requests {instance_group_string} "
+                f"instances. {len(measurements)} measurements were obtained for the model config "
+                f"on CPU with platfrom {platform}. ")
+        else:
+            gpu_dict = self._get_gpu_stats(measurements=measurements)
+            gpu_names = ','.join(list(gpu_dict.keys()))
+            max_memories = ','.join(
+                [str(x) + ' GB' for x in gpu_dict.values()])
+            sentence = (
+                f"The model config {model_config_name} requests {instance_group_string} "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:557]: Can you put {model_config_name} inside double qoutes?

**Review by Tabrizian — CHANGES_REQUESTED** [review]

File: model_analyzer/reports/report_manager.py (hunk 7 above)
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:564]: Yes exactly

File: model_analyzer/reports/report_manager.py (hunk 4 above)
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:548]: currently this displays the instance group string which is of the format {n}/{KIND}, i'll replace the '/' with a space

File: model_analyzer/reports/report_manager.py (hunk 6 above)
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:559]: The only difference would be the GPUs would be a comma separated list followed by their max memories.

**Comment by dzier** [comment]
What is going on with the results? The latency should not sawtooth like that.

**Comment by aramesh7** [comment]
> What is going on with the results? The latency should not sawtooth like that.

I updated the PDF, its not the latency sawtoothing its the throughput. I think it happened because i used a checkpoint with some mixed results.

File: model_analyzer/reports/report_manager.py [hunk 9]
```diff
@@ -458,72 +495,105 @@ def _build_detailed_table(self, model_config_name):
...
+        model_config, measurements = self._detailed_report_data[
+            model_config_name]
+        instance_group_string = model_config.instance_group_string()
+        dynamic_batching_string = model_config.dynamic_batching_string()
+        platform = model_config.get_field('platform')
+
+        # Measurements and GPU info
+        if model_config.cpu_only():
+            sentence = (
+                f"The model config {model_config_name} requests {instance_group_string.replace('/', ' ')} "
```
Comment (Tabrizian) [inline model_analyzer/reports/report_manager.py:557]: Can we change "requests" to "uses"?

**Review by Tabrizian — CHANGES_REQUESTED** [review]

File: model_analyzer/reports/report_manager.py (hunk 9 above)
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py]: Done

**Review by Tabrizian — APPROVED** [review]

**Pull request merged**